### DISCUSSION TITLE
Trouble to define a boundary condition: help please !

### DISCUSSION THREAD

**Discussion by amnchacon** [firedrakeproject/firedrake#3369] [Firedrake support]
Hello everyone, I'm trying to implement this example in Firedrake:
![Captura de tela de 2024-01-27 18-39-00](https://github.com/firedrakeproject/firedrake/assets/146268309/2d7bbaa9-c809-448f-9de5-f1e90c0107bf)

I had already treated this problem with both boundary conditions as Dirichlet boundary conditions, with constant temperature values. Now, my problem is with the flux boundary condition. It suppose that flux boundary conditions are embedden within the variational formulation, but after testing numerical results, these are being inconsistent. My code is running without errors, but the results are not sucessfully. The way I'm trying the variational formulation here in Firedrike is right or wrong? can you help me please?

In my variational formulation via Firedrake I'm writing:

a = -K * grad(T)[0] * v * dx (The part associated to this natural boundary condition)
...
...
L = (...) + h * (T-Ta) * v * dx

The code is running but with strange results, what I should do for traducing correctly to firedrike this flux boundary condition?

Thanks for helping me...

Another thing: such boundary condition -k(dT/dr) = h(To - Ta) contains To, which is the temperature at r = ro, i.e., a temperature on the outside surface, which is a variable.

Besides this, If I would have a square mesh and if I need to define flux boundaries conditions, one with convection and another with a prescript flux value, how to wirte it correctly in the variational formulation via Firedrake?

I have not been able to move forward due to these issues...

My gratitude for responding...

**Comment by colinjcotter** [comment]
If you want to do an integral on the outer surface, you need to use ds, not dx.

(Edit: `ds`, not `dS` the latter is for interior face integrals)

**Reply by amnchacon** [reply]
Thanks. Well, I have written the variational formulation this way (specifically the part associated with this boundary condition):

a = K * grad(T)[0] * v * ds
a += ...
...
L = f * v * dx + h * (T-Ta) * v * ds

The "K * grad(T)[0] * v * ds" and "h * (T-Ta) * v * ds" parts are the new added parts of the variational formulation (firstly I had tested this problem with both dirichlet boundary conditions and it worked very well -I tested as a liner problem, using the "LinearVariationalSolver" function)... after adding them it's appearing this error:

"Provided RHS is not a linear form".

**Reply by colinjcotter** [reply]
A linear form needs to be linear in the test function and independent of trial function. Can you check if this is the case?

Also I think you don’t want to add that surface term to a. It is the term that should be replaced with your flux term. In the derivation you integrate by parts and then substitute the boundary condition.

**Reply by amnchacon** [reply]
I defined like this:
...
T = TrialFunction(V)
v = TestFunction(V)
...
x = SpatialCoordinate(mesh)

rx = Function(V)
rx.interpolate(x[0])

...
...
a = K * grad(T)[0] * v * ds
a = K * rx * inner(grad(T), grad(v)) * dx
a -= ((K+(rx * Var_K)) * grad(T)[0] * v) * dx
L = f * v * dx + h * (T-Ta) * v * ds

solution = Function(V)             
anel_problem = LinearVariationalProblem(a, L, solution, bcs=[bc])
...

based on this variational formulation:
![Captura de tela de 2024-01-27 18-44-13](https://github.com/firedrakeproject/firedrake/assets/146268309/d6cd765b-6658-4dc2-aa79-c93b5029e5bc)

And yes, after integrating by parts we'll have a term which is outside the integral but it's part of the variational equation. That term q (in blue in the image) would be h(T-Ta).... but I'm confused to translate this to the code.

**Reply by colinjcotter** [reply]
OK, your problem is that T is a TrialFunction so it cannot appear in the right hand side.

**Reply by amnchacon** [reply]
Ok, if I delete this term, how can I write correctly this: -kdK/dt = h(To-Ta)... To is a variable, specifically is the outer surface temperature. Different to the inner temperature and atmospheric temperature (Ta), which are prescribed. I deleted "T" of the right hand side appears another error: "exterior_facet integral cannot be used with a TensorProductCell; need to distinguish between vertical and horizontal contributions."

**Reply by colinjcotter** [reply]
OK, take a look at this section for an explanation of surface integrals for extruded meshes.

https://www.firedrakeproject.org/extruded-meshes.html

**Reply by amnchacon** [reply]
Ok Prof. I'll read it carefully and I'll write later. Thanks .

**Reply by amnchacon** [reply]
Well, after reading and carrying out some tests, I applied "ds_t" integration differential instead of "ds" and it worked well, the code ran without errors, but the results are not the expected when these are compared with the analytical solution; this is expected too because writing in the way I wrote is like treating as an integral a term which is deduced after integrating, I mean K * grad(T)[0] * v * ds_t

Back to the variational formulation, after integrate by parts, we have a term which is outside the integral (like the image I sent in comments above) and I think this is the new problem, how to traduce correctly this in the variational formulation? the boundary condition -K*dT/dr = h(To-Ta) has To, and To is a variable we want to know.

**Reply by colinjcotter** [reply]
It sounds like this is a maths problem not a Firedrake problem. Do you have a Robin boundary condition?

**Reply by colinjcotter** [reply]
Here is a 1D example of applying a Robin condition:

$-u_{xx} = f$ with $u_x + u = 666$ at x=0,1

$-\int_0^1 vu_{xx} dx = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [vu_x]_0^1 = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx +[vu]_0^1 = \int_0^1 vf dx-[666v]_0^1$

**Reply by amnchacon** [reply]
Yes Professor, I have a Robin condition. I understood what you did, just one thing: I think in your example the right hand side should be $+[666v]_0^1$:

$\int_0^1 v_xu_x dx - [vu_x]_0^1 = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [v(1)u_x(1) - v(0)u_x(0)] = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [v(1)(666-u(1)) - v(0)(666-u(0))]  = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [666v(1) - v(1)u(1) - 666v(0) + v(0)u(0)]  = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [(666v(1) - 666v(0)) - (v(1)u(1) - v(0)u(0))]  = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [([666v]_0^1) - ([vu]_0^1)]  = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx - [666v]_0^1 + [vu]_0^1  = \int_0^1 vf dx$

$\int_0^1 v_xu_x dx + [vu]_0^1 = \int_0^1 vf dx + [666v]_0^1$

But anyway, in my problem I did the same, well, I had left expressed the term of the derivative. For not showing all the process I did, I'll show one of the terms after integrate by parts, I'd have:

$[KvT_r]_{r_i}^{r_o}$

if my boundary condition at $r=r_o$ is $-KT_r = h(T_o - T_a)$, ($T_r = dT/dr$), applying what you showed in your example, I would have $KT_r = -h(T_o - T_a)$ and:

$[KvT_r]_{r_i}^{r_o}$

$[-h(T_o - Ta)v]_{r_i}^{r_o} = -h(T_o - Ta)v(r_o) + h(T_o - Ta)v(r_i)$

and v(r_i) = 0 because we have a strong boundary condition at $r=r_i$, $T(r_i)=T_w$, so:

$[-h(T_o - Ta)v]_{r_i}^{r_o} = -h(T_o - Ta)v(r_o)$

My question is, how should I write in Firedrake, within the variational formulation this part: $-h(T_o - Ta)v(r_o)$? taking into account that $T_o$ is not a given value, but a variable because is the outside temperature. The first example I did, with both dirichlet boundary conditions, my variational formulation didn't have terms outside the integral, all were within and it worked well. Now I have $-h(T_o - Ta)v(r_o)$.

Or in your example:

$\int_0^1 v_xu_x dx + [vu]_0^1 = \int_0^1 vf dx + [666v]_0^1$

how can we write within the variational formulation these parts: $[vu]_0^1$ ; $[666v]_0^1$ ?
I made myself understood or am I making an error?
